File: TanWeiChit/SkipList/SkipList.java
public class SkipList{
    
    private SkipListNode head;
    private SkipListNode tail;
    private int h;
    public int maxH;
    public static final double P=0.5;
    public static final int Max = 6;
    
    private SkipListNode[] update;
    
    public SkipList(){
        h = 0;
        maxH = 0;
        
        head = new SkipListNode(Integer.MIN_VALUE, 0);
        tail = new SkipListNode(Integer.MAX_VALUE, 0);
        
        head.next[0] = tail;
        update = new SkipListNode[h+1];
    }
    
    private int randheight(){
        int h = 0;
        h = (int)(Math.log(1.-Math.random())/Math.log(1.-P));
        return Math.min(h, Max);
    }
    
    public SkipListNode search(int key){
        SkipListNode p = head;
        for(int i =h; i >= 0; i--){
            while(p.next[i].key < key){
                p=p.next[i];
            }
            p = p.next[0];
            if(p.key == key && p!= tail)
                return p;
            else
                return null;
        }
        return null;
    }
    
    public void insert(int key){
        SkipListNode p = head;
        for(int i = h; i>=0; i--){
            while(p.next[i].key<key)
            {
                p=p.next[i];
            }
            update[i] = p;
        }
        p = p.next[0];
        /*if(p.key == key){
            System.out.println("KEY"+key+"FOUND");
            return;
        }*/
        int newH = randheight();
        if(newH > maxH){
            SkipListNode oldHead = head;
            head = new SkipListNode(Integer.MIN_VALUE, newH);
            for(int i=0; i<=maxH; i++)
                head.next[i] = oldHead.next[i];
            for(int i = maxH+1; i<=newH; i++)
                head.next[i] = tail;
            
            maxH = newH;
            
            for(int i = h; i>=0 && update[i] == oldHead; i--)
                update[i] = head;
            
            SkipListNode[] oldUpdate = update;
            update = new SkipListNode[newH + 1];
            for(int i =0; i<= h; i++)
                update[i] = oldUpdate[i];
        }
        if(newH > h){
            for(int i = h+1; i <= newH; i++)
                update[i] = head;
            h = newH;
        }
        
        p = new SkipListNode(key, newH);
        
        for(int i=0; i<= newH; i++)
        {
            p.next[i] = update[i].next[i];
            update[i].next[i] =p;
        }
    }
    
    public void delete(int key){
        SkipListNode p =head;
        SkipListNode[] update = new SkipListNode[h+1];
        
        for(int i =h; i>=0; i--)
        {
            while(p.next[i].key < key)
            {
                p= p.next[i];
            }
            update[i] = p;
        }
        p = p.next[0];
        
        if(p.key != key)
        {
            System.out.println("KEY " + key+ " NOT FOUND");
            return;
        }
        for(int i =0; i< p.next.length; i++)
        {
            update[i].next[i] = update[i].next[i].next[i];
        }
        
        while(h>=0 && head.next[h] == tail)
            h--;
        
        if(4*h <= maxH){
            maxH = 2*h;
            SkipListNode oldHead = head;
            head = new SkipListNode(Integer.MIN_VALUE, maxH);
            for(int i=0; i<=maxH; i++)
                head.next[i] = oldHead.next[i];
            
            update = new SkipListNode[maxH+1];
        }
    }
        public void print(){
            for(int i = h; i>=0; i--){
                SkipListNode p = head.next[i];
                SkipListNode q = head.next[0];
                while(p!=tail){
                    if(q == p){
                        System.out.print(p.key+" ");
                        p = p.next[i];
                    }
                    else System.out.print("   ");
                    q = q.next[0];
                }
                System.out.println();
            }
            System.out.println();
        }
    }
                    
            
        
        
        

File: Dong Fei/Homework 2/skiplist.cpp
//
//  main.cpp
//  Alg_Project2-2
//
//  Created by D&F on 12/11/13.
//  Copyright (c) 2013 D&F. All rights reserved.
//

//Skip list implementation

#include <iostream>
//#include <math.h>

#define valueType char
//#define MaxLevel 4

using namespace std;

// SNode class
class SNode
{
public:
    //int level;
    valueType value;
    int Max_Level;
    SNode **forward;
    SNode(int size){ forward= new SNode *[size];};   //constructor of class
    ~SNode(){delete []forward;}
    
    friend class Slist;
};

//Slist class
class Slist
{
public:
    Slist(valueType tKey, int MaxL=4);                    //
    ~Slist();
    
    int MaxLevel;
    int tailKey;
    int levels;                         //current node's level
    
    SNode *head;                        //head node/pointer
    SNode *tail;                        //tail pointer
    SNode **next;                       //poniter array
    
    int level();                        //generate random level number
    
    bool Search(valueType v);
    SNode *Slist_Search(valueType v);
    
    Slist &Slist_Insert(valueType v);
    
    Slist &Slist_Delete(valueType v);
    
    
};


//Slist constructor

Slist::Slist(valueType tKey, int MaxL)                 //constructor
{
    tailKey=tKey;                                           //tail key set as a big value
    MaxLevel=MaxL;                    //calculate MaxLevel
    
    head=new SNode (MaxLevel+1);
    tail=new SNode (0);
    next=new SNode *[MaxLevel+1];
    
    tail->value=tKey;
    
    
    for (int i=0; i<=MaxLevel;i++)
        head->forward[i]=tail;

}


//Slist destructor
Slist::~Slist()
{
    SNode *p;
    
    while(head!=tail)
    {
        p=head->forward[0];
        delete head;
        head=p;
    }
    delete  tail;
    delete []next;
}


//Search function
//return find result
bool Slist::Search(valueType v)
{
    valueType e;
    if(v>=tailKey)
     return false;
    
    SNode *p=head;
    for(int i=levels; i>=0; i--)                //find the predescessor of  v
    {
        while(p->forward[i]->value<v)
            p=p->forward[i];
    }
    e=p->forward[0]->value;
    
    if(e==v)
        cout<<"Congratulations! find the result!" <<endl;
    else
        cout<<"Sorry, not found!" <<endl;
    
    return (e==v);                          //if the value of  next node equals to v, return 1.
    
}

//Slist_Search function : return SNode

SNode *Slist::Slist_Search(valueType v)  //find the predescessor of v and return the node.
{
    
    if(v>=tailKey)
        return NULL;
    
    SNode *p=head;
    for(int i=levels; i>=0; i--)                //find the predescessor node of  v
    {
        while(p->forward[i]->value<v)
            p=p->forward[i];
        next[i]=p;                              //record every matching node
    }
    
    return (p->forward[0]);
    
}


//level function
// to randomly generate a level number of SNode
int Slist::level()
{
    int lev=0;
    lev=rand()%MaxLevel;
    return lev>0?lev:lev+1;
}

// Slist_Insertion function : insert new node
Slist &Slist::Slist_Insert(valueType v)
{
    if(v>=tailKey)
    {
        cout<<"Inserting number is too big, please change a small one! "<< endl;
        return *this;
    }

    SNode *p=Slist_Search(v);
    if(p->value==v)
    {
        cout<<"Value repete, input another one!"<<endl;
        return *this;
    }
    
    int lev=level();//randomly generate a level number for new node
    
    if(lev>levels)
    {
        lev=++levels;       //fix level to levels +1 if lev >levels
        next[lev]=head;
    }
    
    SNode *y=new SNode (lev+1);
    y->value=v;
    for(int i=0; i<lev; i++)
    {
        y->forward[i]=next[i]->forward[i];
        next[i]->forward[i]=y;
    }
    return *this;
    
}

//Slist_Delete function : delete node with value v
Slist &Slist::Slist_Delete(valueType v)
{
    
    SNode *p=Slist_Search(v);
    if(v>=tailKey || p->value!=v)
    {
        cout<< "no such value!"<<endl;
        return *this;
    }
    
    for(int i=0; i<levels && next[i]->forward[i]==p; i++)       //directly delete node
    {
        next[i]->forward[i]=p->forward[i];
    }
    
    while(levels>0 && head->forward[levels]==tail)
        levels--;
    valueType e=p->value;
    cout<< "delete value:  "<< e<<endl;
    delete p;
    return *this;
}


//Main Function
int main()
{

//if valueType is int type
    
 /*   Slist SL(10000,4);                  //10000,tail node key, set as large as possible
    SL.Slist_Insert(100);
    SL.Slist_Insert(10);
    SL.Slist_Insert(17);
    SL.Slist_Insert(46);
    SL.Slist_Insert(9);
    SL.Slist_Insert(19);
    SL.Slist_Insert(88);
    SL.Search(10);
    SL.Slist_Delete(10);
    SL.Search(10);
  */

//if  valueType is char type
    valueType v1,v2;


    Slist SL('~',4);

    SL.Slist_Insert('C');
    SL.Slist_Insert('B');
    SL.Slist_Insert('h');
    SL.Slist_Insert('X');
    SL.Slist_Insert('j');
    SL.Slist_Insert('K');
//    SL.Slist_Insert('p');
    SL.Slist_Insert('Z');
    
    cout<<"Please input the key to search(A-Z, or a-z): \n";
    cin>>v1;
    
    SL.Search(v1);
    cout<<""<<endl;
    cout<<"Please input the key to delete(A-Z, or a-z): \n";
    cin>>v2;
    
    SL.Slist_Delete(v2);

    cout<< " Search after delete: "<<v2<<endl;
    SL.Search(v2);
    return 0;
}

